lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -843,6 +878,98 @@ void FunctionEntrypointAnalyzer::interproceduralPropagation() {
+    if (It != ABIRegisterMap.end()) {
+      revng_assert(It->second != State::Count && It->second != State::Invalid);
+      return It->second;
+    }
+  }
+
+  return std::nullopt;
+}
+
+void FEA::initializeMapForDeductions(FunctionSummary &Summary,
+                                     abi::RegisterState::Map &Map) {
+  auto Arch = model::ABI::getArchitecture(Binary->DefaultABI);
+  revng_assert(Arch == Binary->Architecture);
+
+  for (const auto &Reg : model::Architecture::registers(Arch)) {
+    const auto &ArgRegisters = Summary.ABIResults.ArgumentsRegisters;
+    const auto &RVRegisters = Summary.ABIResults.FinalReturnValuesRegisters;
+
+    if (auto MaybeState = tryGetRegisterState(Reg, ArgRegisters))
+      Map[Reg].IsUsedForPassingArguments = *MaybeState;
+
+    if (auto MaybeState = tryGetRegisterState(Reg, RVRegisters))
+      Map[Reg].IsUsedForReturningValues = *MaybeState;
+  }
+}
+
+void FunctionEntrypointAnalyzer::applyABIDeductions() {
+  using namespace abi;
+  using State = RegisterState::Values;
+
+  if (ABIEnforcement.getNumOccurrences() == 0
+      || ABIEnforcement == NoABIEnforcement)
+    return;
+
+  for (const model::Function &Function : Binary->Functions) {
+    auto &Summary = Oracle.at(Function.Entry);
+
+    RegisterState::Map StateMap(Binary->Architecture);
+    initializeMapForDeductions(Summary, StateMap);
+
+    bool EnforceABIConformance = ABIEnforcement == FullABIEnforcement;
+    std::optional<abi::RegisterState::Map>
+      ResultMap = EnforceABIConformance ?
+                    enforceRegisterStateDeductions(StateMap,
+                                                   Binary->DefaultABI) :
+                    tryApplyRegisterStateDeductions(StateMap,
+                                                    Binary->DefaultABI);
+
+    if (!ResultMap.has_value())
+      continue;
+
+    for (const auto &[Register, State] : *ResultMap) {
+      llvm::StringRef Name = model::Register::getCSVName(Register);
+      if (llvm::GlobalVariable *CSV = M.getGlobalVariable(Name, true)) {
+        auto MaybeArg = State.IsUsedForPassingArguments;
+        auto MaybeRV = State.IsUsedForReturningValues;
+
+        if (RegisterState::isYesOrDead(MaybeArg))
+          if (Summary.ABIResults.ArgumentsRegisters.count(CSV) != 0)
+            Summary.ABIResults.ArgumentsRegisters[CSV] = State::Yes;
+
+        if (MaybeArg == State::No || MaybeArg == State::NoOrDead)
+          if (Summary.ABIResults.ArgumentsRegisters.count(CSV) != 0)
+            Summary.ABIResults.ArgumentsRegisters[CSV] = State::No;
+
+        if (RegisterState::isYesOrDead(MaybeRV))
+          if (Summary.ABIResults.FinalReturnValuesRegisters.count(CSV) != 0)
+            Summary.ABIResults.FinalReturnValuesRegisters[CSV] = State::Yes;
+
+        if (MaybeRV == State::No || MaybeRV == State::NoOrDead)
+          if (Summary.ABIResults.FinalReturnValuesRegisters.count(CSV) != 0)
+            Summary.ABIResults.FinalReturnValuesRegisters[CSV] = State::No;
```

Comment:
If `MaybeArg == State::YesOrDead`, don't we want the final State to be promoted to `Yes`?

Reply:
Why? This phase should just let the ABI deductions refine results. Why do some additional arbitrary refinement here?

What's the final decision (argument or not) it's a competence of the phase in which we commit to the model. I don't see why we should do this here.

Reply:
I'm just not sure why we cannot simply promote `YesOrDead` to `Yes` directly at this point while we are at it, but sure, that's fine.

Reply:
Because there might be further refinement passes one day that decide differently.
I see no point in doing any extra processing w.r.t. to what ABI deductions decided...